vllm/v1/core/kv_cache_utils.py

```diff
@@ -541,48 +571,113 @@ def _get_kv_cache_config_uniform_type(vllm_config: VllmConfig,
     TODO: support hybrid models with more than one type of KV cache.
 
     Args:
         vllm_config: The global VllmConfig
-        kv_cache_specs: The kv cache specs of the model
+        kv_cache_spec: The kv cache spec of each attention layer in the model
         available_memory: Memory available for KV cache in bytes.
 
     Returns:
         The generated KVCacheConfigs
     """
-    # Use the max number of layers to conservatively determine
-    # the number of blocks.
-    num_layers = max(len(kv_cache_spec) for kv_cache_spec in kv_cache_specs)
-    kv_cache_configs = []
-    for kv_cache_spec in kv_cache_specs:
-        check_enough_kv_cache_memory(vllm_config, kv_cache_spec,
-                                     available_memory)
-        if is_kv_cache_type_uniform(kv_cache_spec):
-            # KV cache of all layers are the same, which is true for
-            # most models. Allocate the same amount of memory for
-            # each layer.
-            kv_cache_configs.append(
-                _get_kv_cache_config_uniform_type(vllm_config, kv_cache_spec,
-                                                  available_memory,
-                                                  num_layers))
-        else:
-            raise NotImplementedError
+    check_enough_kv_cache_memory(vllm_config, kv_cache_spec, available_memory)
+    make_uniform_kv_cache_type(kv_cache_spec)
+    if is_kv_cache_type_uniform(kv_cache_spec):
+        # KV cache of all layers are the same, which is true for
+        # most models. Allocate the same amount of memory for
+        # each layer.
+        return _get_kv_cache_config_uniform_type(vllm_config, kv_cache_spec,
+                                                 available_memory)
+
+    raise NotImplementedError
+
+
+def make_kv_cache_configs_consistent(kv_cache_configs: List[KVCacheConfig]):
+    """
+    Make the KV cache configurations for each worker consistent, so that all 
+    workers can be controlled by the same KVCacheManager.
+    This function verifies that the virtual layers of each worker are the same,
+    and changes the num_blocks of each worker to the smallest among all workers.
+    
+    Args:
+        kv_cache_configs: The KV cache configurations for each worker. Will be
+            in-place modified to make them consistent.
+    """
+
+    # Sort the virtual layers by the type_id of the KV cache spec.
+    # This can avoid the inconsistency caused by the order of virtual layers.
+    for kv_cache_config in kv_cache_configs:
+        kv_cache_config.virtual_layers.sort(
+            key=lambda x: x.kv_cache_spec.type_id)
+
+    # Verify that the virtual layers of each rank are the same.
+    for kv_cache_config in kv_cache_configs[1:]:
+        for virtual_layer1, virtual_layer2 in zip(
+                kv_cache_configs[0].virtual_layers,
+                kv_cache_config.virtual_layers):
+            assert virtual_layer1.kv_cache_spec == virtual_layer2.kv_cache_spec
```

Comment:
Q: Will pipeline parallelism fail this assert for some models?

Reply:
Yes, it will fail when different stages have different type of layers. For hybrid models, just throw an error as the first step. For non-hybrid models, the assert won't fail.

This function is introduced in https://github.com/vllm-project/vllm/pull/14079/files, should we discuss it there?